dev/release/bump_version.sh

```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+set -eu
+
+SOURCE_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+SOURCE_TOP_DIR="$(cd "${SOURCE_DIR}/../../" && pwd)"
+
+if [ "$#" -ne 1 ]; then
+  echo "Usage: $0 <version>"
+  echo " e.g.: $0 19.0.1"
+  echo " e.g.: $0 20.0.0-SNAPSHOT"
+  exit 1
+fi
+
+version=$1
+
+cd "${SOURCE_TOP_DIR}"
+
+git switch -c "bump-version-${version}"
+mvn versions:set "-DnewVersion=${version}" -DprocessAllModules -DgenerateBackupPoms=false
+case "${version}" in
+*-SNAPSHOT)
+  tag=main
+  ;;
+*)
+  tag=v${version}
+  ;;
+esac
+mvn versions:set-scm-tag "-DnewTag=${tag}" -DgenerateBackupPoms=false -pl :arrow-java-root
```

Comment:
Why do we need set-scm-tag for `SNAPSHOT` version?

Reply:
Because we don't create a branch for release in the proposed workflow.

My understanding is that this updates `project.scm.tag`: https://github.com/apache/arrow-java/blob/4a535af9e92b11fc618f15e3dd9cdfad8c554ca2/pom.xml#L84

`pom.xml` in the main branch uses `vX.Y.Z` while we're releasing. We need to back it to `main` after we released `X.Y.Z`.

Do we want to create a release branch? If we have a release branch, we don't need to back it to `main` because `pom.xml` in the `main` branch always uses `main` for `project.scm.tag`. `pom.xml` in release branches use `vX.Y.Z` for `project.scm.tag`.

(If we create a release branch, release process will be a bit complex.)

Reply:
If we want to follow up with a 19.0.1 release to fix issues in 19.0.0, does it mean that we have to create the RC from main branch too?

Reply:
Yes.

FYI: apache/arrow-julia and apache/arrow-go use this style.

If we have enough resource to maintain multiple major releases (e.g. 18.Y.Z and 19.Y.Z), we can create a release branch for each major release. But I think that we don't have enough resource for now...

Reply:
Agreed. Let's not over-complicate the script for now.